Occlusion Rendering

RGregat opened the pull request:
* Implemented the Depth API to seamlessly blend the virtual with the real world. 
* Different modes are implemented and can be used as follows

### DepthMode
> The DepthMode is automatically set based on the Session config.

**NO_DEPTH**
The {@link Session} is not configured to use the Depth-API
This is the default value

**DEPTH**
The {@link Session} is configured to use the DepthMode AUTOMATIC

**RAW_DEPTH**
The {@link Session} is configured to use the DepthMode RAW_DEPTH_ONLY

### DepthOcclusionMode
> The DepthOcclusionMode can be set by the user.

**DEPTH_OCCLUSION_DISABLED**
Use this value if the standard camera material should be applied to the CameraStream Renderable even if the {@link Session} configuration has set the DepthMode to {@link Config.DepthMode#AUTOMATIC} or {@link Config.DepthMode#RAW_DEPTH_ONLY}. This Option is useful, if you want to use the DepthImage or RawDepthImage or just the DepthPoints without the occlusion effect.
This is the default value

**DEPTH_OCCLUSION_ENABLED**
Set the occlusion material. If the {@link Session} is not configured properly the standard camera material is used. Valid {@link Session} configuration for the DepthMode are {@link Config.DepthMode#AUTOMATIC} and {@link Config.DepthMode#RAW_DEPTH_ONLY}.

### Examples

```java
@Override
public void onSessionConfiguration(Session session, Config config) {
    if (session.isDepthModeSupported(Config.DepthMode.AUTOMATIC))
        config.setDepthMode(Config.DepthMode.AUTOMATIC);
    config.setUpdateMode(Config.UpdateMode.LATEST_CAMERA_IMAGE);
}

@Override
public void onViewCreated(ArFragment arFragment, ArSceneView arSceneView) {
    // By default he occlusion is disabled even if the DepthMode is set to AUTOMATIC or RAW_DEPTH_ONLY
    // Not every Depth use case requires the occlusion feature!
    arSceneView
        .getCameraStream()
        .setDepthOcclusionMode(CameraStream.DepthOcclusionMode
            .DEPTH_OCCLUSION_ENABLED); // Available modes: DEPTH_OCCLUSION_DISABLED, DEPTH_OCCLUSION_ENABLED
}
```

![Occlusion_AUTOMATIC](https://user-images.githubusercontent.com/3974162/122067720-1a300100-cdf4-11eb-8af1-8969c0df1f3e.jpg)

```java
@Override
public void onSessionConfiguration(Session session, Config config) {
    if (session.isDepthModeSupported(Config.DepthMode.RAW_DEPTH_ONLY))
        config.setDepthMode(Config.DepthMode.RAW_DEPTH_ONLY);
    config.setUpdateMode(Config.UpdateMode.LATEST_CAMERA_IMAGE);
}

@Override
public void onViewCreated(ArFragment arFragment, ArSceneView arSceneView) {
    // By default he occlusion is disabled even if the DepthMode is set to AUTOMATIC or RAW_DEPTH_ONLY
    // Not every Depth use case requires the occlusion feature!
    arSceneView
        .getCameraStream()
        .setDepthOcclusionMode(CameraStream.DepthOcclusionMode
            .DEPTH_OCCLUSION_ENABLED); // Available modes: DEPTH_OCCLUSION_DISABLED, DEPTH_OCCLUSION_ENABLED
    }
```

![Occlusion_RAW_DEPTH_ONLY](https://user-images.githubusercontent.com/3974162/122067801-27e58680-cdf4-11eb-8568-92124a23d336.jpg)

```java
@Override
public void onSessionConfiguration(Session session, Config config) {
    if (session.isDepthModeSupported(Config.DepthMode.DISABLED))
        config.setDepthMode(Config.DepthMode.DISABLED);
    config.setUpdateMode(Config.UpdateMode.LATEST_CAMERA_IMAGE);
}
```

![Occlusion_DISABLED](https://user-images.githubusercontent.com/3974162/122067844-32078500-cdf4-11eb-8ed4-c667dea6933e.jpg)

Any of the following combinations results into the disabled Occlusion use case

```java
@Override
public void onSessionConfiguration(Session session, Config config) {
    if (session.isDepthModeSupported(Config.DepthMode.AUTOMATIC))
        config.setDepthMode(Config.DepthMode.DISABLED);
   config.setUpdateMode(Config.UpdateMode.LATEST_CAMERA_IMAGE);
}
```
```java
@Override
public void onSessionConfiguration(Session session, Config config) {
    if (session.isDepthModeSupported(Config.DepthMode.RAW_DEPTH_ONLY))
        config.setDepthMode(Config.DepthMode.DISABLED);
    config.setUpdateMode(Config.UpdateMode.LATEST_CAMERA_IMAGE);
}
```

If the user wants to activate the DepthMode by setting it to AUTOMATIC or RAW_DEPTH_ONLY but is not interested about the occlusion of virtual objects behind real world objects he can set the DepthOcclusionMode value to DEPTH_OCCLUSION_DISABLED. 

```java
@Override
public void onViewCreated(ArFragment arFragment, ArSceneView arSceneView) {
    // Currently, the tone-mapping should be changed to FILMIC
    // because with other tone-mapping operators except LINEAR
    // the inverseTonemapSRGB function in the materials can produce incorrect results.
    // The LINEAR tone-mapping cannot be used together with the inverseTonemapSRGB function.
    Renderer renderer = arSceneView.getRenderer();

    if (renderer != null) {
        renderer.getFilamentView().setColorGrading(
            new ColorGrading.Builder()
                .toneMapping(ColorGrading.ToneMapping.FILMIC)
                .build(EngineInstance.getEngine().getFilamentEngine())
        );
    }

    arSceneView
        .getCameraStream()
        .setDepthOcclusionMode(CameraStream.DepthOcclusionMode
            .DEPTH_OCCLUSION_ENABLED); // Available modes: DEPTH_OCCLUSION_DISABLED, DEPTH_OCCLUSION_ENABLED
    }
```

grassydragon commented on core/assets-sources/materials/ar_environment_material_depth_test.mat:
```diff
@@ -0,0 +1,41 @@
+material {
```
We need to remove this file.

RGregat replied:
done

RGregat commented:
Hmm well, I guess a toggle is needed to be able to not use the occlusion even if the DepthMode is enabled!

grassydragon commented on core/src/main/java/com/google/ar/sceneform/SceneView.java:
```diff
@@ -22,6 +22,8 @@
 import com.google.ar.sceneform.utilities.MovingAverageMillisecondsTracker;
```
Can these changes be removed?

grassydragon replied:
I meant that were some new lines and a new import but now it is okay.

grassydragon commented on core/src/main/java/com/google/ar/sceneform/rendering/DepthTexture.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * <pre>
+ *     The DepthTexture class holds a special Texture to store
+ *     information from a DepthImage to realize the occlusion of
```
"implement the occlusion"?

RGregat replied:
What you mean with "implement the occlusion"?

grassydragon replied:
realize -> implement

RGregat replied:
done

grassydragon commented on core/src/main/java/com/google/ar/sceneform/rendering/DepthTexture.java:
```diff
@@ -0,0 +1,121 @@
+
+import androidx.annotation.Nullable;
+
+import com.google.android.filament.Texture;
```
This import or imports for each Texture can be removed.

RGregat replied:
done

grassydragon commented on core/src/main/java/com/google/ar/sceneform/rendering/MaterialParameters.java:
```diff
@@ -1,12 +1,15 @@
 package com.google.ar.sceneform.rendering;
 
+import android.util.Log;
```
Can the Log import be removed?

RGregat replied:
done

grassydragon commented on core/src/main/java/com/google/ar/sceneform/rendering/MaterialParameters.java:
```diff
@@ -252,6 +268,16 @@ void applyTo(MaterialInstance materialInstance) {
+    com.google.android.filament.Material material = materialInstance.getMaterial();
+
+    if(material.hasParameter(name)) {
+      Optional
```
Maybe it is better to use `if` in Java?

RGregat replied:
Optional is available since Java 8. To be fair, it is a personal thing of mine to use Optional. But I have no problem to replace it with an if statement.

grassydragon commented on core/src/main/java/com/google/ar/sceneform/rendering/Renderer.java:
```diff
@@ -287,6 +287,7 @@ public void render(long frameTimeNanos, boolean debugEnabled) {
 
         // Render the scene, unless the renderer wants to skip the frame.
         // This means you are sending frames too quickly to the GPU
+        //renderer.beginFrame(swapChainLocal, frameTimeNanos);
```
This commented call should be removed.

RGregat replied:
done

grassydragon commented on samples/gltf/src/main/java/com/google/ar/sceneform/samples/gltf/MainActivity.java:
```diff
@@ -34,7 +36,7 @@
         FragmentOnAttachListener,
         BaseArFragment.OnTapArPlaneListener,
-        ArFragment.OnViewCreatedListener {
+        ArFragment.OnViewCreatedListener, BaseArFragment.OnSessionConfigurationListener {
```
This should be put on a new line together with `BaseArFragment.OnTapArPlaneListener`.

RGregat replied:
done

grassydragon commented on core/assets-sources/materials/ar_environment_material_flat.mat:
```diff
@@ -0,0 +1,35 @@
+material {
```
Where is this material used?

RGregat replied:
That is the standard camera material if the DepthMode is disabled

RGregat commented:
I added a toggle to switch the used Material. It is possible to use the standard camera material even when the DepthMode is set to DEPTH or RAW_DEPTH_ONLY. 
I implemented it in a rush and it needs some refinement, but thats something for tomorrow.

```java
@Override
public void onViewCreated(ArFragment arFragment, ArSceneView arSceneView) {
    ...
    arSceneView
       .getCameraStream() 
       .setDepthModeUsage(CameraStream.DepthModeUsage.DEPTH_MODE_ENABLED);
}
```

RGregat commented:
Updated the initial PR Comment

ThomasGorisse commented on samples/depth/src/main/res/layout/view_tiger_card.xml:
```diff
@@ -0,0 +1,10 @@
+    android:gravity="center"
+    android:orientation="vertical"
+    android:padding="6dp"
+    android:text="Tiger"
```
```suggestion
    android:text="Giant Panda"
```

ThomasGorisse commented on samples/depth/src/main/java/com/google/ar/sceneform/samples/depth/MainActivity.java:
```diff
@@ -0,0 +1,167 @@
+        if (session.isDepthModeSupported(Config.DepthMode.AUTOMATIC))
+            config.setDepthMode(Config.DepthMode.AUTOMATIC);
+
+        config.setUpdateMode(Config.UpdateMode.LATEST_CAMERA_IMAGE);
```
Why? Isn't it the default mode?
```suggestion
```

ThomasGorisse commented on samples/depth/src/main/res/drawable/rounded_bg.xml:
```diff
@@ -0,0 +1,21 @@
+   limitations under the License.
+-->
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <solid android:color="#E61976d2"/>
```
```suggestion
    <solid android:color="#E6FFFFFF"/>
```

ThomasGorisse commented on samples/depth/src/main/java/com/google/ar/sceneform/samples/depth/MainActivity.java:
```diff
@@ -0,0 +1,167 @@
+            );
+        }
+
+        arSceneView
+                .getCameraStream()
+                .setDepthOcclusionMode(CameraStream.DepthOcclusionMode
+                        .DEPTH_OCCLUSION_ENABLED); // Available modes: DEPTH_OCCLUSION_DISABLED, DEPTH_OCCLUSION_ENABLED
```
```suggestion
        // Available modes: DEPTH_OCCLUSION_DISABLED, DEPTH_OCCLUSION_ENABLED
        arSceneView.getCameraStream()
            .setDepthOcclusionMode(CameraStream.DepthOcclusionMode.DEPTH_OCCLUSION_ENABLED);
```
